packages/cli/commands/project/cloneApp.js

```diff
@@ -0,0 +1,171 @@
+  const accountConfig = getAccountConfig(accountId);
+  const accountName = uiAccountDescription(accountId);
+
+  trackCommandUsage('clone-app', {}, accountId);
+
+  if (!isAppDeveloperAccount(accountConfig)) {
+    uiLine();
+    logger.error(i18n(`${i18nKey}.errors.invalidAccountTypeTitle`));
+    logger.log(
+      i18n(`${i18nKey}.errors.invalidAccountTypeDescription`, {
+        useCommand: uiCommandReference('hs accounts use'),
+        authCommand: uiCommandReference('hs auth'),
+      })
+    );
+    uiLine();
+    process.exit(EXIT_CODES.ERROR);
+  }
+
+  const { appId } =
+    'appId' in options
+      ? options
+      : await selectPublicAppPrompt({
+          accountId,
+          accountName,
+          options,
+          migrateApp: false,
+        });
+
+  const publicApps = await fetchPublicAppOptions(accountId, accountName);
+  const selectedApp = publicApps.find(a => a.id === appId);
+  if (!selectedApp) {
+    logger.error(i18n(`${i18nKey}.errors.invalidAppId`, { appId }));
+    process.exit(EXIT_CODES.ERROR);
+  }
+
+  const { location } =
+    'location' in options
+      ? options
+      : await promptUser({
+          name: 'location',
+          message: i18n(`${i18nKey}.enterLocation`),
+          default: path.resolve(getCwd(), selectedApp.name),
+          validate: input => {
+            if (!input) {
+              return i18n(`${i18nKey}.errors.locationRequired`);
+            }
+            if (fs.existsSync(input)) {
+              return i18n(`${i18nKey}.errors.invalidLocation`);
+            }
+            if (!isValidPath(input)) {
+              return i18n(`${i18nKey}.errors.invalidCharacters`);
+            }
+            return true;
+          },
+        });
+
+  try {
+    SpinniesManager.init();
+
+    SpinniesManager.add('cloneApp', {
+      text: i18n(`${i18nKey}.cloneStatus.inProgress`),
+    });
+
+    const cloneResponse = await cloneApp(accountId, appId);
+    const { exportId } = cloneResponse;
+    const pollResponse = await poll(checkCloneStatus, accountId, exportId);
+    const { status } = pollResponse;
+    if (status === 'SUCCESS') {
+      const absoluteDestPath = path.resolve(getCwd(), location);
+      const zippedProject = await downloadClonedProject(accountId, exportId);
+
+      await extractZipArchive(
+        zippedProject,
+        selectedApp.name,
+        path.resolve(absoluteDestPath),
+        { includesRootDir: true, hideLogs: true }
+      );
+
+      SpinniesManager.succeed('cloneApp', {
+        text: i18n(`${i18nKey}.cloneStatus.done`),
+        succeedColor: 'white',
+      });
+      logger.log('');
+      uiLine();
+      logger.success(i18n(`${i18nKey}.cloneStatus.success`, { location }));
+      logger.log('');
+      process.exit(EXIT_CODES.SUCCESS);
+    }
+  } catch (error) {
+    SpinniesManager.fail('cloneApp', {
+      text: i18n(`${i18nKey}.cloneStatus.failure`),
+      failColor: 'white',
+    });
+    if (error.errors) {
```

Comment:
When does `error.errors` exist? Should we have handling for these in our apiErrors util in local-dev-lib? I have a similar gut feeling here that shrinking the size of the try/catch blocks will make error handling more straightforward. Some of the errors thrown will be filesystem errors and some will be api errors. We're calling `logApiErrorInstance` in the catch block, but what happens if it was the `extractZipArchive ` util that failed?

Reply:
The migrations endpoints (that includes migration and cloning) return a response object that looks like this:

```
{
id: 12234,
status: 'PENDING',
errors?: [Error, Error] //I know this is bad pseudo code, apologies
}
```

I don't believe that we need to account for this in our error handling in local-dev-lib, given that this is a special case (Projects can have multiple reasons why migration/cloning fails). I've accounted for other error types (as in the error by `extractZipArchive` with the conditional statement. I'm going to leave as is.

Reply:
Added a note directly in the code about this: https://github.com/HubSpot/hubspot-cli/pull/1085/commits/80217fc0b34dc2fdf1b979b78f2c19967105fbbe